### PR TITLE
Option to Read Excel without a pop-up Window

### PR DESCRIPTION
**Pull request by dimven**
Please Note:
1. Before submitting the PR, please review [How to Contribute to Dynamo](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md)
2. Dynamo Team will meet 1x a month to review PRs found on Github (Issues will be handled separately)
3. PRs will be reviewed from oldest to newest
4. If a reviewed PR requires changes by the owner, the owner of the PR has 30 days to respond. If the PR has seen no activity by the next session, it will be either closed by the team or depending on its utility will be taken over by someone on the team
5. PRs should use either Dynamo's default PR template or [one of these other template options](https://github.com/DynamoDS/Dynamo/wiki/Choosing-a-Pull-Request-Template) in order to be considered for review.
6. PRs that do not have one of the Dynamo PR templates completely filled out with all declarations satisfied will not be reviewed by the Dynamo team.
7. PRs made to the `DynamoRevit` repo will need to be cherry-picked into all the DynamoRevit Release branches that Dynamo supports. Contributors will be responsible for cherry-picking their reviewed commits to the other branches after a `LGTM` label is added to the PR.

### Purpose

After working with Excel for the whole day, seeing the same windows pop up for a bajilion times starts to get annoying. (and I may not be the only one https://github.com/DynamoDS/Dynamo/issues/7921 ) This PR adds the option to read sheets without any pop-ups. I'd love to hear any ideas on how to improve on this approach.

The default behavior of the node will remain the same, the inner workings of the reading process do not change. Dynamo still handles the life-cycle of the excel application and these changes should not affect that in any way. 

The only major change in the logic is the need to close the workbook after every hidden read. This is necessary to make sure that we don't have any invisible ghost workbooks bogging down the excel process. If you're reading multiple sheets from the same workbook, this could possibly affect performance negatively. If a user sees any performance degradation or his existing workflow is compromised, he can always just revert to the default behavior.

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers
@aparajit-pratap

I noticed you're reordering the class structure and the library for 2.0. This is using the new "ImportExcel" method. Not sure how you want to expose the input descriptions.

### FYIs

@kronz

### CODE REVIEW

File: src/Libraries/DSOffice/Excel.cs [hunk 1]
```diff
@@ -598,6 +609,15 @@ internal WorkBook(string filePath)
             else
                 wb = ExcelInterop.App.Workbooks.Add();
         }
+        
+        /// <summary>
+        /// Helper method for reading workbooks with a disabled visibility.
+        /// </summary>
+        public void CloseHidden()
```
Comment (aparajit-pratap) [inline src/Libraries/DSOffice/Excel.cs]: This method can be private

File: src/Libraries/DSOffice/Excel.cs [hunk 2]
```diff
@@ -269,17 +269,28 @@ public static WorkBook SaveAsExcelWorkbook(WorkBook workbook, string filename)
...
         /// <search>office,excel,spreadsheet,ifequalreturnindex</search>
         [IsVisibleInDynamoLibrary(false)]
-        public static object[][] ReadFromFile(FileInfo file, string sheetName, bool readAsStrings = false)
+        public static object[][] ReadFromFile(FileInfo file, string sheetName, bool readAsStrings = false, bool visible = true)
         {
-            WorkBook wb = WorkBook.ReadExcelFile(file.FullName);
+        	object[][] data;
+        	
+        	if(!visible)
+        		ExcelInterop.ShowOnStartup = false;
```
Comment (aparajit-pratap) [inline src/Libraries/DSOffice/Excel.cs]: Our coding convention recommends putting statements in `if-else` blocks inside parentheses even if they are single statements, like:
```
if(!visible)
{
    ExcelInterop.ShowOnStartup = false;
}
```

**Comment by dimven** [comment]
I've updated the code as per the standards. I can promote the "CloseHidden" method to internal. I can't make it private, because I have to call it on the workbook instance.

If I enable the visibility for the old method, it works as expected (testing with 1.31 stable). 
Hidden:

dynamosandbox_2017-08-16_19-45-35

and visible:

**Comment by aparajit-pratap** [comment]
@dimven thanks! We'll test your changes and consider it for submission. One thing I notice is that the `visible` flag on the node makes sense only when the `ImportExcel` node is the first Excel node to execute in the graph because if another excel node is run before it, excel would be open anyway? @kronz this PR looks like a much needed improvement we must consider to be merged.

**Comment by jnealb** [comment]
@aparajit-pratap Is there a corresponding JIRA task for this work?

**Comment by kronz** [comment]
The functionality looks good, and I'm glad it is being done in consideration of #7921. One question: "visible" seems a little opaque, any other more descriptive but terse term we can use?  @Racel any ideas?

**Comment by aparajit-pratap** [comment]
@jnealb no, there is no task for this; it was done for us by one of our users for free :)

**Comment by jnealb** [comment]
@aparajit-pratap QNTM-1649 so we can track it.

**Comment by aparajit-pratap** [comment]
Thanks @jnealb

**Comment by jnealb** [comment]
[~pratapa] Two questions for you:
1. Why have the default be "visible" when the more frequent desired behavior is invisible? The user has to hook up a Boolean node, which should not be necessary. If the default was invisible then we could introduce a "ShowExcel" node whose only purpose is to pop it up.
2. The Data.ExportExcel node needs this "visible" port for symmetry. Why release a matched pair of functions that are asymmetric in this regard, is there a reason?
And finally one speculative thought: this may improve our Excel tests, which suffer from false positives (if Export also supported an invsible mode).

**Comment by aparajit-pratap** [comment]
@jnealb the default is `visible` because that was the previous behaviour, excel would be shown by default. To your point about `ShowExcel` and `ExportExcel` nodes, you may have a point; I would defer to @kronz or @Racel. Btw I think `showExcel` might be a better name for the port instead of `visible`.

**Comment by kronz** [comment]
I like "Show Excel" as a port name, I agree that it should ber Show Excel = true by default as 
a) it is more understandable with Positive statement (Show) and Positive state (Yes!)
b) it is more comprehensible for first time users (do something, see a result),
c) presevation of old behavior.  

The problem I have with a dedicated Show Excel node is that it would represent something (display of data) that new users would not necessarily even knew that they could ask for.

**Comment by Racel** [comment]
I know this is already merged. But, here are my 2 cents.
- I don't think an explicit "Show Excel" node is necessary. 
- I have been thinking that there should be some way for users to know if a parameter is optional or has a default. Maybe the easiest thing to do is add an asterisks after the portname? I haven't given this much thought, but would like to start a discussion
- I also would move away from using the term visible as it is a term that is overloaded in Dynamo. I think `showExcel` sounds good to me, but maybe we can even think of other terms like `excelSilent`, `runSilent`,  or something like that
- the default of this port should be true to preserve existing behavior
- I agree with @jnealb - `ExportExcel` should be symmetric as using it will also bring up the excel window.

**Pull request merged**